homeassistant/components/recorder/history/modern.py

```diff
@@ -559,40 +617,77 @@ def _get_start_time_state_for_entities_stmt(
     metadata_ids: list[int],
     no_attributes: bool,
     include_last_changed: bool,
+    slow_dependant_subquery: bool,
 ) -> Select:
     """Baked query to get states for specific entities."""
-    # This query is the result of significant research in
-    # https://github.com/home-assistant/core/issues/132865
-    # A reverse index scan with a limit 1 is the fastest way to get the
-    # last state change before a specific point in time for all supported
-    # databases. Since all databases support this query as a join
-    # condition we can use it as a subquery to get the last state change
-    # before a specific point in time for all entities.
-    stmt = (
-        _stmt_and_join_attributes_for_start_state(
-            no_attributes, include_last_changed, False
+    if slow_dependant_subquery:
+        # Simple group-by for MySQL, must use less
+        # than 1000 metadata_ids in the IN clause for MySQL
+        # or it will optimize poorly.
+        most_recent_states_for_entities_by_date = (
+            select(
+                States.metadata_id.label("max_metadata_id"),
+                func.max(States.last_updated_ts).label("max_last_updated"),
+            )
+            .filter(
+                (States.last_updated_ts < epoch_time)
+                & States.metadata_id.in_(metadata_ids)
+            )
+            .group_by(States.metadata_id)
+            .subquery()
         )
-        .select_from(StatesMeta)
-        .join(
-            States,
-            and_(
-                States.last_updated_ts
-                == (
-                    select(States.last_updated_ts)
-                    .where(
-                        (StatesMeta.metadata_id == States.metadata_id)
-                        & (States.last_updated_ts < epoch_time)
+        stmt = (
+            _stmt_and_join_attributes_for_start_state(
+                no_attributes, include_last_changed, False
```

Comment:
Maybe not for this PR, but this kind of signature makes for hard to read code:
```py
def _stmt_and_join_attributes(
    no_attributes: bool,
    include_last_changed: bool,
    include_last_reported: bool,
) -> Select:
```

When called like this:
```py
            _stmt_and_join_attributes_for_start_state(
                no_attributes, include_last_changed, False
            )
```
We should either make the bool arguments kwarg-only, to enforce readable code, or at least not pass unnamed bool constants.

Reply:
I usually switch it when it gets to 4-5 args unless there is a performance concern because we execute it 10000+/s.  In this case we definitely don't have a performance concern and I'm happy to do it for 3 as well

changed in 8f2eb8c3ac1